### PR TITLE
Add new BBEs related to objects

### PR DESCRIPTION
**Pull request by prakanth97**
## Purpose

Add following BBEs,

- Object types
- Object type inclusion
- Distinct object types

### CODE REVIEW

File: examples/distinct-object-types/distinct_object_types.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+## Distinct object types
```
Comment (prakanth97) [inline examples/distinct-object-types/distinct_object_types.md:1]:
```suggestion
# Distinct object types
```

File: examples/object-types/object_types.md [hunk 2]
```diff
@@ -0,0 +1,7 @@
+## Object types
+
+An `object type` is a type definition without the implementation. It is similar to an interface, as defined in the `java` programming language.
```
Comment (lasinicl) [inline examples/object-types/object_types.md:3]:
```suggestion
An object type is a type definition without any implementation. It is similar to an interface, as defined in the Java programming language.
```

File: examples/object-types/object_types.out [hunk 3]
```diff
@@ -0,0 +1,2 @@
+bal run object_types.bal
```
Comment (lasinicl) [inline examples/object-types/object_types.out:1]: Should be 
```suggestion
$ bal run object_types.bal
```
Let's fix this in other `out` files too

File: examples/object-types/object_types.metatags [hunk 4]
```diff
@@ -0,0 +1,2 @@
+description: This bbe demonstrates object types in ballerina
```
Comment (lasinicl) [inline examples/object-types/object_types.metatags:1]:
```suggestion
description: This BBE demonstrates object types in Ballerina
```
Let's fix in other places too

File: examples/object-types/object_types.bal [hunk 5]
```diff
@@ -0,0 +1,27 @@
+import ballerina/io;
+
+// This is the definition of `Hashable` object type
```
Comment (lasinicl) [inline examples/object-types/object_types.bal:3]:
```suggestion
// `Hashable` object type with a method called `hash( )` that returns an integer.
```

File: examples/object-types/object_types.bal [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+type Hashable object {
+    function hash() returns int;
+};
+
+function h() returns any {
+    var obj = object {
+        // This is the implementation of method `hash()` that defined in `Hashable` object type,
+        // like implementing an interface
```
Comment (lasinicl) [inline examples/object-types/object_types.bal:11]:
```suggestion
       // Implements the `hash()` method defined in `Hashable` object type.
```

File: examples/object-types/object_types.bal [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+        function hash() returns int {
+            return 42;
+        }
+
+        // It can have other methods
```
Comment (lasinicl) [inline examples/object-types/object_types.bal:16]:
```suggestion
        // The object can have other methods.
```

File: examples/object-types/object_types.bal [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+        function zero() returns int {
+            return 0;
+        }
+    };
+    return obj;
+}
+
+public function main() {
+    // output is true
```
Comment (lasinicl) [inline examples/object-types/object_types.bal:25]:
```suggestion
    // The returned object matches the pattern of the `Hashable` object type, which contains a `hash()` method returning an integer.
```

File: examples/object-type-inclusion/object_type_inclusion.md [hunk 9]
```diff
@@ -0,0 +1,7 @@
+## Object type inclusion
+
+Type inclusion enables to create an object type that includes another object type, such that one interface extends another interface and, creates a class that includes a type, like the class implementing the interface. Using the `*T` notation, you can include the object type to another object type or another class.
```
Comment (lasinicl) [inline examples/object-type-inclusion/object_type_inclusion.md:3]: > Type inclusion enables to create an object type that includes another object type, such that one interface extends another interface and, creates a class that includes a type, like the class implementing the interface.

This sentence seems a bit too long. Shall we break it into maybe 2 sentences? It'll be easier to understand then.

Comment (lasinicl) [reply examples/object-type-inclusion/object_type_inclusion.md:3]: > Using the `*T` notation, you can include the object type to another object type or another class.

Shall we just say 
> You can include object types using the `*T` syntax.

here. What we have now, seems to be repeating the same thing in the previous sentence.

File: examples/object-type-inclusion/object_type_inclusion.bal [hunk 10]
```diff
@@ -0,0 +1,43 @@
+import ballerina/io;
+
+type Cloneable object {
+    function clone() returns Cloneable;
+};
+
+// `Shape` object type extends the `Clonable` object type, like interface extends another interface
+type Shape object {
+    *Cloneable;
```
Comment (lasinicl) [inline examples/object-type-inclusion/object_type_inclusion.bal:9]:
```suggestion
type Shape object {
     // `Cloneable` object type is included as a part of the interface for the `Shape` object type. 
    *Cloneable;
```

File: examples/object-type-inclusion/object_type_inclusion.bal [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+
```
Comment (lasinicl) [inline examples/object-type-inclusion/object_type_inclusion.bal:10]:
```suggestion

        // `draw()` is a part of `Shape`'s own type. Function `clone()` is also included from the `Cloneable` type.
```

File: examples/object-type-inclusion/object_type_inclusion.bal [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+// `Shape` object type extends the `Clonable` object type, like interface extends another interface
+type Shape object {
+    *Cloneable;
+    
+    function draw() returns string;
+};
+
+// Circle class implements `Shape`, like implementing an interface
+class Circle {
+    *Shape;
```
Comment (lasinicl) [inline examples/object-type-inclusion/object_type_inclusion.bal:16]:
```suggestion
class Circle {
    // The class `Circle` includes the `Shape` object type.
    *Shape;
```

File: examples/object-type-inclusion/object_type_inclusion.bal [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+
+    int radius;
+
+    function init(int radius) {
+        self.radius = radius;
+    }
+
+    // Returning `Circle` is valid, since the `Circle` type becomes a subtype of `Cloneable` type
+    // once it includes the `Cloneable` object type
```
Comment (lasinicl) [inline examples/object-type-inclusion/object_type_inclusion.bal:25]:
```suggestion
    // once it includes the `Cloneable` object type.
```

File: examples/object-type-inclusion/object_type_inclusion.bal [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+    function draw() returns string {
+        return string `circle:${self.radius}`;
+    }
+}
+
+public function main() {
+    Circle circle = new Circle(5);
+    Circle circleClone = circle.clone();
+    
+    // `circle` and `circleClone` have different memory location, but identical objects
```
Comment (lasinicl) [inline examples/object-type-inclusion/object_type_inclusion.bal:39]: Shall we refrain from adding java like comments  :grimacing:

File: examples/object-type-inclusion/object_type_inclusion.md (hunk 9 above)
Comment (lasinicl) [reply examples/object-type-inclusion/object_type_inclusion.md:3]: Shall we also include

> The implementation of the object type within the class that includes the type is checked at the compile time. This provides interface inheritance. Ballerina does not support implementation inheritance.

File: examples/distinct-object-types/distinct_object_types.md [hunk 15]
```diff
@@ -0,0 +1,9 @@
...
+
+The `distinct` keyword creates distinct object types. This concept allows to define a type with nominal typing within a structured type system. This is useful while interacting with the external world through API interfaces like `GraphQL`, you may want to leverage nominal typing via this distinct typing feature of Ballerina.
```
Comment (lasinicl) [inline examples/distinct-object-types/distinct_object_types.md:3]:
```suggestion
Using the `distinct` keyword in the type definition creates distinct object types. This concept allows to define a type with nominal typing within a structured type system. This is useful while interacting with the external world through API interfaces like `GraphQL`, you may want to leverage nominal typing via this distinct typing feature of Ballerina.
```

File: examples/distinct-object-types/distinct_object_types.bal [hunk 16]
```diff
@@ -0,0 +1,37 @@
+import ballerina/io;
+
```
Comment (lasinicl) [inline examples/distinct-object-types/distinct_object_types.bal:2]: Shall we add a comment for this too?

File: examples/distinct-object-types/distinct_object_types.bal [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+function desc(Employee employee) returns string {
+    // The `is` operator can be used to distinguish distinct subtypes.
+    return employee is Engineer ? "Engineer" : "Manager";
+}
+
+public function main() {
+    Employee employee = new Engineer("James");
+    io:println(desc(employee));
+
+    // `Engineer employee2 = new Manager("clark");` gives compilation error
```
Comment (lasinicl) [inline examples/distinct-object-types/distinct_object_types.bal:36]: We don't usually add compilation error-related scenarios in BBEs

File: examples/object-type-inclusion/object_type_inclusion.md (hunk 9 above)
Comment (prakanth97) [reply examples/object-type-inclusion/object_type_inclusion.md]: Updated 9d5526315210351e8e92f1c0d36238e5ae574152

File: examples/object-type-inclusion/object_type_inclusion.bal [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+        return string `circle:${self.radius}`;
+    }
+}
+
+public function main() {
+    Circle circle = new Circle(5);
+    Circle circleClone = circle.clone();
+    
+    io:println(circle === circleClone);
+    io:println(circle.draw());
```
Comment (SasinduDilshara) [inline examples/object-type-inclusion/object_type_inclusion.bal:43]: Isn't it better if we move this near to the declaration?.
I am suggesting to move it to after L39

**Review by lasinicl — APPROVED** [review]

File: examples/distinct-object-types/distinct_object_types.bal [hunk 19]
```diff
@@ -0,0 +1,36 @@
+import ballerina/io;
+
+// `Person` object type that contains a string field called `name`.
```
Comment (praneesha) [inline examples/distinct-object-types/distinct_object_types.bal:3]:
```suggestion
// The `Person` object type that contains a string field called `name`.
```

File: examples/distinct-object-types/distinct_object_types.bal [hunk 20]
```diff
@@ -0,0 +1,36 @@
...
+type Person distinct object {
+    public string name;
+};
+
+// `Engineer` and `Manager` classes are structurally the same but introducing the `distinct` keyword
```
Comment (praneesha) [inline examples/distinct-object-types/distinct_object_types.bal:8]:
```suggestion
// The `Engineer` and `Manager` classes are structurally the same but introducing the `distinct` keyword
```

File: examples/distinct-object-types/distinct_object_types.md [hunk 21]
```diff
@@ -0,0 +1,9 @@
+# Distinct object types
+
+Using the `distinct` keyword in the type definition creates distinct object types. This concept allows to define a type with nominal typing within a structured type system. This is useful while interacting with the external world through API interfaces like `GraphQL`, you may want to leverage nominal typing via this distinct typing feature of Ballerina.
```
Comment (praneesha) [inline examples/distinct-object-types/distinct_object_types.md:3]:
```suggestion
Using the `distinct` keyword in the type definition creates distinct object types. This concept allows to define a type with nominal typing within a structured type system. This is useful while interacting with the external world through API interfaces like `GraphQL`. You may want to leverage nominal typing via this distinct typing feature of Ballerina.
```

File: examples/distinct-object-types/distinct_object_types.metatags [hunk 22]
```diff
@@ -0,0 +1,2 @@
+description: This BBE demonstrates distinct object types in Ballerina
```
Comment (praneesha) [inline examples/distinct-object-types/distinct_object_types.metatags:1]:
```suggestion
description: This BBE demonstrates distinct object types in Ballerina.
```

File: examples/object-type-inclusion/object_type_inclusion.bal [hunk 23]
```diff
@@ -0,0 +1,46 @@
+import ballerina/io;
+
+type Cloneable object {
+    function clone() returns Cloneable;
+};
+
+type Shape object {
+    // `Cloneable` object type is included as a part of the interface for the `Shape` object type.
```
Comment (praneesha) [inline examples/object-type-inclusion/object_type_inclusion.bal:8]:
```suggestion
    // The `Cloneable` object type is included as a part of the interface of the `Shape` object type.
```

File: examples/object-type-inclusion/object_type_inclusion.bal [hunk 24]
```diff
@@ -0,0 +1,46 @@
...
+    *Cloneable;
+
+    // `draw()` is a part of `Shape`'s own type. Function `clone()` is also included from the
```
Comment (praneesha) [inline examples/object-type-inclusion/object_type_inclusion.bal:11]:
```suggestion
    // `draw()` is a part of `Shape`'s own type. The `clone()` function is also included from the
```

File: examples/object-type-inclusion/object_type_inclusion.bal [hunk 25]
```diff
@@ -0,0 +1,46 @@
...
+    // `Cloneable` type.
+    function draw() returns string;
+};
+
+class Circle {
+    // The class `Circle` includes the `Shape` object type. Therefore it has to implement both
```
Comment (praneesha) [inline examples/object-type-inclusion/object_type_inclusion.bal:17]:
```suggestion
    // The `Circle` class includes the `Shape` object type. Therefore, it has to implement both 
```

File: examples/object-type-inclusion/object_type_inclusion.bal [hunk 26]
```diff
@@ -0,0 +1,46 @@
...
+    // the `clone()` and `draw()` methods.
+    *Shape;
+
+    int radius;
+
+    function init(int radius) {
+        self.radius = radius;
+    }
+
+    // Returning `Circle` is valid, since the `Circle` type becomes a subtype of `Cloneable` type
```
Comment (praneesha) [inline examples/object-type-inclusion/object_type_inclusion.bal:27]:
```suggestion
    // Returning `Circle` is valid as the `Circle` type becomes a subtype of the `Cloneable` type
```

File: examples/object-type-inclusion/object_type_inclusion.md [hunk 27]
```diff
@@ -0,0 +1,9 @@
+# Object type inclusion
+
+Object type inclusion enables two things. First, it allows create an object type that includes another object type, such that one interface extends another interface. Second, it allows create a class that includes a type, like the class implementing the interface. You can include object types using the *T syntax.
```
Comment (praneesha) [inline examples/object-type-inclusion/object_type_inclusion.md:3]:
```suggestion
Object type inclusion enables two things. Firstly, it allows the creation of an object type that includes another object type such that one interface extends another interface. Secondly, it allows creating a class that includes a type like the class implementing the interface. You can include object types using the `*T` syntax.
```

File: examples/object-type-inclusion/object_type_inclusion.metatags [hunk 28]
```diff
@@ -0,0 +1,2 @@
+description: This BBE demonstrates object type inclusion in Ballerina
```
Comment (praneesha) [inline examples/object-type-inclusion/object_type_inclusion.metatags:1]:
```suggestion
description: This BBE demonstrates object-type inclusion in Ballerina.
```

File: examples/object-types/object_types.bal [hunk 29]
```diff
@@ -0,0 +1,27 @@
+import ballerina/io;
+
+// `Hashable` object type with a method called `hash( )` that returns an integer.
```
Comment (praneesha) [inline examples/object-types/object_types.bal:3]:
```suggestion
// The `Hashable` object type with a method called `hash( )`, which returns an integer.
```

File: examples/object-types/object_types.bal [hunk 30]
```diff
@@ -0,0 +1,27 @@
...
+type Hashable object {
+    function hash() returns int;
+};
+
+function h() returns any {
+    var obj = object {
+        // Implements the `hash()` method defined in `Hashable` object type.
```
Comment (praneesha) [inline examples/object-types/object_types.bal:10]:
```suggestion
        // Implements the `hash()` method defined in the `Hashable` object type.
```

File: examples/object-types/object_types.md [hunk 31]
```diff
@@ -0,0 +1,7 @@
+# Object types
+
+An object type is a type definition without any implementation. It is similar to an interface, as defined in the Java programming language. Object typing is structural, and an object type looks like a pattern that the object must match.
```
Comment (praneesha) [inline examples/object-types/object_types.md:3]:
```suggestion
An object type is a type definition without any implementation. It is similar to an interface as defined in the Java programming language. Object typing is structural and an object type looks like a pattern that the object must match.
```

File: examples/object-types/object_types.metatags [hunk 32]
```diff
@@ -0,0 +1,2 @@
+description: This BBE demonstrates object types in Ballerina
```
Comment (praneesha) [inline examples/object-types/object_types.metatags:1]:
```suggestion
description: This BBE demonstrates object types in Ballerina.
```

**Review by praneesha — APPROVED** [review]

File: examples/distinct-object-types/distinct_object_types.bal [hunk 33]
```diff
@@ -0,0 +1,36 @@
+import ballerina/io;
+
+// The `Person` object type that contains a string field called `name`.
```
Comment (pcnfernando) [inline examples/distinct-object-types/distinct_object_types.bal:3]: Shall we remove `The` from the type comments

File: examples/distinct-object-types/distinct_object_types.bal [hunk 34]
```diff
@@ -0,0 +1,36 @@
...
+    
+    function init(string name) {
+        self.name = name;
+    }
+}
+
+type Employee Engineer|Manager;
+
+function desc(Employee employee) returns string {
+    // The `is` operator can be used to distinguish distinct subtypes.
```
Comment (pcnfernando) [inline examples/distinct-object-types/distinct_object_types.bal:29]: Here as well

Comment (prakanth97) [reply examples/distinct-object-types/distinct_object_types.bal]: Updated d8c6a9e496c62f91e5499baaa4413757c01d639c

File: examples/distinct-object-types/distinct_object_types.bal (hunk 33 above)
Comment (prakanth97) [reply examples/distinct-object-types/distinct_object_types.bal]: Updated d8c6a9e496c62f91e5499baaa4413757c01d639c

File: examples/distinct-object-types/distinct_object_types.bal [hunk 35]
```diff
@@ -0,0 +1,36 @@
...
+    
+    function init(string name) {
+        self.name = name;
+    }
+}
+
+type Employee Engineer|Manager;
+
+function desc(Employee employee) returns string {
+    // `is` operator can be used to distinguish distinct subtypes.
```
Comment (praneesha) [inline examples/distinct-object-types/distinct_object_types.bal:29]:
```suggestion
    // The `is` operator can be used to distinguish distinct subtypes.
```

File: examples/object-types/object_types.bal [hunk 36]
```diff
@@ -0,0 +1,27 @@
+import ballerina/io;
+
+// `Hashable` object type with a method called `hash()`, which returns an integer.
```
Comment (praneesha) [inline examples/object-types/object_types.bal:3]:
```suggestion
// The `Hashable` object type with a method called `hash()`, which returns an integer.
```

**Review by praneesha — APPROVED** [review]

File: examples/distinct-object-types/distinct_object_types.bal (hunk 33 above)
Comment (praneesha) [reply examples/distinct-object-types/distinct_object_types.bal]: Better to add it for grammatical correctness if we don't have a strong reason to remove it.

**Pull request merged**